### PR TITLE
Parsimmon: strict null checks and typed languages

### PR DESCRIPTION
**Pull request by bvanreeven**
Please fill in this template.

- [x] Use a meaningful title for the pull request. Include the name of the package modified.
- [x] Test the change in your own code. (Compile and run.)
- [x] Follow the advice from the [readme](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#make-a-pull-request).
- [x] Avoid [common mistakes](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#common-mistakes).
- [x] Run `npm run lint package-name` (or `tsc` if no `tslint.json` is present).

Select one of these and delete the others:

If changing an existing definition:
- [x] Provide a URL to documentation or source code which provides context for the suggested changes: https://github.com/jneen/parsimmon/blob/master/API.md#parsimmoncreatelanguageparsers
- [x] Increase the version number in the header if appropriate.
- [x] If you are making substantial changes, consider adding a `tslint.json` containing `{ "extends": "dtslint/dt.json" }`.

### CODE REVIEW

**Review by bcherny — APPROVED** [review]
Awesome! Thanks for the contribution!

**Comment by bcherny** [comment]
👍

File: types/parsimmon/parsimmon-tests.ts [hunk 1]
```diff
@@ -191,3 +191,55 @@ strArrPar = P.sepBy1(P.string('foo'), P.string('bar'));
...
+// --  --  --  --  --  --  --  --  --  --  --  --  --
+
+let language: Language;
+
+language = P.createLanguage({
+	SomeRule: r => P.alt(P.string(""), r.AnotherRule),
+	AnotherRule: () => P.string(""),
+});
+
+// $ExpectType Parser<any>
```
Comment (bcherny) [inline types/parsimmon/parsimmon-tests.ts]: @bvanreeven Is there a way to better type these? Ie. this rule should ideally be `Parser<string>`.

Comment (bvanreeven) [reply types/parsimmon/parsimmon-tests.ts]: As far as I can tell, this is not possible without introducing additional type parameters: one for the `rules` parameter, and one for the return type.

The return type of `createLanguage` should be an object that has the same properties as the `rules` parameter object, but the types of those properties should match the return types of the functions that are passed as properties of the `rules` parameter object. If that makes sense.

I tried an approach that has only one type parameter and returns a mapped type of the parameter, but I got stuck here:

```typescript
function createLanguage<TRule extends Rule>(rules: TRule): { [K in keyof TRule]: <return type of>TRule[K] };
```

As far as I know, it is currently not possible to statically determine the return type of a function type.

Comment (bvanreeven) [reply types/parsimmon/parsimmon-tests.ts]: @bcherny I added some documentation on how to use the type parameter in `createLanguage` to make a language specification typed.

**Pull request merged**